File: python/feedpulse/parser.py
"""Feed data normalization and parsing.

This module provides parsers for different feed types:
- HackerNews API
- Reddit JSON
- Lobsters JSON
- GitHub Events API

Each parser normalizes feed-specific formats into FeedItem objects.
"""

import sys
import json
from typing import Any, Optional

from .models import FeedItem
from .utils import coerce_to_string, parse_timestamp
from .exceptions import MalformedJSONError, UnexpectedStructureError


def parse_hackernews(data: Any, source: str) -> list[FeedItem]:
    """Parse HackerNews API response.
    
    Supports two formats:
    1. Top stories API: array of story IDs [12345, 23456, ...]
    2. Algolia search API: {hits: [{objectID, title, url, ...}, ...]}
    
    Args:
        data: Parsed JSON data
        source: Source feed name
        
    Returns:
        List of FeedItem objects
    """
    items: list[FeedItem] = []
    
    try:
        # Check if it's the Algolia format with 'hits'
        if isinstance(data, dict) and 'hits' in data:
            hits = data.get('hits', [])
            if not isinstance(hits, list):
                print(f"Warning: {source}: 'hits' must be a list", file=sys.stderr)
                return items
            
            for i, hit in enumerate(hits):
                try:
                    if not isinstance(hit, dict):
                        continue
                    
                    # Get title
                    title = hit.get('title')
                    if not title:
                        continue
                    title = coerce_to_string(title)
                    if not title:
                        continue
                    
                    # Get URL (or generate from objectID)
                    url = hit.get('url')
                    if not url:
                        story_id = hit.get('objectID') or hit.get('story_id')
                        if story_id:
                            url = f"https://news.ycombinator.com/item?id={story_id}"
                        else:
                            continue
                    
                    url = coerce_to_string(url)
                    if not url:
                        continue
                    
                    # Parse timestamp
                    timestamp = None
                    created_at = hit.get('created_at_i') or hit.get('created_at')
                    if created_at:
                        timestamp = parse_timestamp(created_at)
                    
                    item = FeedItem.create(
                        title=title,
                        url=url,
                        source=source,
                        timestamp=timestamp,
                        raw_data=hit
                    )
                    items.append(item)
                    
                except Exception as e:
                    print(f"Warning: {source}: item {i}: failed to parse: {e}", file=sys.stderr)
                    continue
            
            return items
        
        # Original format: list of story IDs
        if not isinstance(data, list):
            print(f"Warning: {source}: expected list or dict with 'hits', got {type(data).__name__}", file=sys.stderr)
            return items
        
        for i, story_id in enumerate(data[:30]):  # Limit to top 30
            try:
                # Validate ID
                if not isinstance(story_id, int):
                    print(f"Warning: {source}: item {i}: expected int ID, got {type(story_id).__name__}", 
                          file=sys.stderr)
                    continue
                
                title = f"HN Story {story_id}"
                url = f"https://news.ycombinator.com/item?id={story_id}"
                
                item = FeedItem.create(
                    title=title,
                    url=url,
                    source=source,
                    raw_data={'id': story_id}
                )
                items.append(item)
                
            except Exception as e:
                print(f"Warning: {source}: item {i}: failed to parse: {e}", file=sys.stderr)
                continue
    
    except Exception as e:
        print(f"Error: {source}: failed to parse HackerNews data: {e}", file=sys.stderr)
    
    return items


def parse_github(data: Any, source: str) -> list[FeedItem]:
    """
    Parse GitHub API response.
    Extract from items[]: full_name → title, html_url → url, 
    topics → tags, updated_at → timestamp
    """
    items: list[FeedItem] = []
    
    try:
        if not isinstance(data, dict):
            print(f"Warning: {source}: expected dict, got {type(data).__name__}", file=sys.stderr)
            return items
        
        github_items = data.get('items', [])
        if not isinstance(github_items, list):
            print(f"Warning: {source}: 'items' must be a list", file=sys.stderr)
            return items
        
        for i, item_data in enumerate(github_items):
            try:
                if not isinstance(item_data, dict):
                    print(f"Warning: {source}: item {i}: must be a dict", file=sys.stderr)
                    continue
                
                # Required: full_name → title
                title = item_data.get('full_name')
                if not title:
                    title = item_data.get('name')  # Fallback
                if not title:
                    print(f"Warning: {source}: item {i}: missing 'full_name' or 'name'", file=sys.stderr)
                    continue
                
                title = coerce_to_string(title)
                if not title:
                    print(f"Warning: {source}: item {i}: could not convert title to string", file=sys.stderr)
                    continue
                
                # Required: html_url → url
                url = item_data.get('html_url')
                if not url:
                    print(f"Warning: {source}: item {i}: missing 'html_url'", file=sys.stderr)
                    continue
                
                url = coerce_to_string(url)
                if not url:
                    print(f"Warning: {source}: item {i}: could not convert url to string", file=sys.stderr)
                    continue
                
                # Optional: topics → tags
                tags = item_data.get('topics', [])
                if not isinstance(tags, list):
                    tags = []
                else:
                    tags = [str(t) for t in tags if t]
                
                # Optional: updated_at → timestamp
                timestamp = parse_timestamp(item_data.get('updated_at'))
                
                item = FeedItem.create(
                    title=title,
                    url=url,
                    source=source,
                    timestamp=timestamp,
                    tags=tags,
                    raw_data=item_data
                )
                items.append(item)
                
            except Exception as e:
                print(f"Warning: {source}: item {i}: failed to parse: {e}", file=sys.stderr)
                continue
    
    except Exception as e:
        print(f"Error: {source}: failed to parse GitHub data: {e}", file=sys.stderr)
    
    return items


def parse_reddit(data: Any, source: str) -> list[FeedItem]:
    """
    Parse Reddit API response.
    Extract from data.children[].data: title → title, url → url,
    created_utc → timestamp, link_flair_text → tags
    """
    items: list[FeedItem] = []
    
    try:
        if not isinstance(data, dict):
            print(f"Warning: {source}: expected dict, got {type(data).__name__}", file=sys.stderr)
            return items
        
        data_obj = data.get('data', {})
        if not isinstance(data_obj, dict):
            print(f"Warning: {source}: 'data' must be a dict", file=sys.stderr)
            return items
        
        children = data_obj.get('children', [])
        if not isinstance(children, list):
            print(f"Warning: {source}: 'children' must be a list", file=sys.stderr)
            return items
        
        for i, child in enumerate(children):
            try:
                if not isinstance(child, dict):
                    print(f"Warning: {source}: item {i}: child must be a dict", file=sys.stderr)
                    continue
                
                item_data = child.get('data', {})
                if not isinstance(item_data, dict):
                    print(f"Warning: {source}: item {i}: child.data must be a dict", file=sys.stderr)
                    continue
                
                # Required: title
                title = item_data.get('title')
                if not title:
                    print(f"Warning: {source}: item {i}: missing 'title'", file=sys.stderr)
                    continue
                
                title = coerce_to_string(title)
                if not title:
                    print(f"Warning: {source}: item {i}: could not convert title to string", file=sys.stderr)
                    continue
                
                # Required: url
                url = item_data.get('url')
                if not url:
                    print(f"Warning: {source}: item {i}: missing 'url'", file=sys.stderr)
                    continue
                
                url = coerce_to_string(url)
                if not url:
                    print(f"Warning: {source}: item {i}: could not convert url to string", file=sys.stderr)
                    continue
                
                # Optional: created_utc → timestamp (unix timestamp)
                timestamp = parse_timestamp(item_data.get('created_utc'))
                
                # Optional: link_flair_text → tags (single item)
                tags = []
                flair = item_data.get('link_flair_text')
                if flair:
                    flair_str = coerce_to_string(flair)
                    if flair_str:
                        tags = [flair_str]
                
                item = FeedItem.create(
                    title=title,
                    url=url,
                    source=source,
                    timestamp=timestamp,
                    tags=tags,
                    raw_data=item_data
                )
                items.append(item)
                
            except Exception as e:
                print(f"Warning: {source}: item {i}: failed to parse: {e}", file=sys.stderr)
                continue
    
    except Exception as e:
        print(f"Error: {source}: failed to parse Reddit data: {e}", file=sys.stderr)
    
    return items


def parse_lobsters(data: Any, source: str) -> list[FeedItem]:
    """
    Parse Lobsters API response.
    Extract from root array: title → title, url → url (fallback to comments_url),
    created_at → timestamp, tags → tags
    """
    items: list[FeedItem] = []
    
    try:
        if not isinstance(data, list):
            print(f"Warning: {source}: expected list, got {type(data).__name__}", file=sys.stderr)
            return items
        
        for i, item_data in enumerate(data):
            try:
                if not isinstance(item_data, dict):
                    print(f"Warning: {source}: item {i}: must be a dict", file=sys.stderr)
                    continue
                
                # Required: title
                title = item_data.get('title')
                if not title:
                    print(f"Warning: {source}: item {i}: missing 'title'", file=sys.stderr)
                    continue
                
                title = coerce_to_string(title)
                if not title:
                    print(f"Warning: {source}: item {i}: could not convert title to string", file=sys.stderr)
                    continue
                
                # Required: url (fallback to comments_url)
                url = item_data.get('url') or item_data.get('comments_url')
                if not url:
                    print(f"Warning: {source}: item {i}: missing 'url' and 'comments_url'", file=sys.stderr)
                    continue
                
                url = coerce_to_string(url)
                if not url:
                    print(f"Warning: {source}: item {i}: could not convert url to string", file=sys.stderr)
                    continue
                
                # Optional: created_at → timestamp
                timestamp = parse_timestamp(item_data.get('created_at'))
                
                # Optional: tags
                tags = item_data.get('tags', [])
                if not isinstance(tags, list):
                    tags = []
                else:
                    tags = [str(t) for t in tags if t]
                
                item = FeedItem.create(
                    title=title,
                    url=url,
                    source=source,
                    timestamp=timestamp,
                    tags=tags,
                    raw_data=item_data
                )
                items.append(item)
                
            except Exception as e:
                print(f"Warning: {source}: item {i}: failed to parse: {e}", file=sys.stderr)
                continue
    
    except Exception as e:
        print(f"Error: {source}: failed to parse Lobsters data: {e}", file=sys.stderr)
    
    return items


def parse_feed(response_body: str, feed_type: str, source: str) -> list[FeedItem]:
    """Parse feed response based on feed_type.
    
    Supports direct feed types (hackernews, reddit, github, lobsters)
    and auto-detection mode (json).
    
    Args:
        response_body: Raw feed response (JSON string)
        feed_type: Feed type ('hackernews', 'reddit', 'github', 'lobsters', or 'json')
        source: Source feed name (for logging)
        
    Returns:
        List of normalized FeedItem objects
    """
    # Parse JSON
    try:
        data = json.loads(response_body)
    except json.JSONDecodeError as e:
        print(f"Error: {source}: malformed JSON response: {e}", file=sys.stderr)
        return []
    except Exception as e:
        print(f"Error: {source}: failed to parse response: {e}", file=sys.stderr)
        return []
    
    # Route to appropriate parser based on explicit feed_type
    if feed_type == 'hackernews':
        return parse_hackernews(data, source)
    elif feed_type == 'reddit':
        return parse_reddit(data, source)
    elif feed_type == 'github':
        return parse_github(data, source)
    elif feed_type == 'lobsters':
        return parse_lobsters(data, source)
    elif feed_type == 'json':
        # Auto-detect based on structure
        if isinstance(data, list):
            # Could be HackerNews (list of ints) or Lobsters (list of dicts)
            if data and isinstance(data[0], int):
                return parse_hackernews(data, source)
            else:
                return parse_lobsters(data, source)
        elif isinstance(data, dict):
            if 'items' in data:
                return parse_github(data, source)
            elif 'data' in data and isinstance(data.get('data'), dict):
                return parse_reddit(data, source)
            elif 'hits' in data:
                return parse_hackernews(data, source)
        
        print(f"Warning: {source}: unable to auto-detect JSON feed structure", file=sys.stderr)
        return []
    elif feed_type in ('rss', 'atom'):
        # Not implemented for this experiment
        print(f"Warning: {source}: RSS/Atom parsing not implemented", file=sys.stderr)
        return []
    else:
        print(f"Warning: {source}: unknown feed_type '{feed_type}'", file=sys.stderr)
        return []

File: go/internal/parser/parser.go
package parser

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"strconv"
	"time"

	"feedpulse/internal/storage"
)

// ParseResult represents the result of parsing a feed
type ParseResult struct {
	Items  []storage.FeedItem
	Errors []string
}

// Parser handles feed parsing and normalization
type Parser struct{}

// NewParser creates a new parser instance
func NewParser() *Parser {
	return &Parser{}
}

// Parse parses raw feed data and returns normalized items
func (p *Parser) Parse(source string, feedType string, data []byte) ParseResult {
	var result ParseResult

	switch feedType {
	case "json":
		result = p.parseJSON(source, data)
	case "rss":
		result.Errors = append(result.Errors, "RSS parsing not implemented in this version")
	case "atom":
		result.Errors = append(result.Errors, "Atom parsing not implemented in this version")
	default:
		result.Errors = append(result.Errors, fmt.Sprintf("unknown feed type: %s", feedType))
	}

	return result
}

// parseJSON parses JSON feeds (HackerNews, GitHub, Reddit, Lobsters)
func (p *Parser) parseJSON(source string, data []byte) ParseResult {
	var result ParseResult

	// Try to parse as generic JSON first
	var rawJSON interface{}
	if err := json.Unmarshal(data, &rawJSON); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("malformed JSON: %v", err))
		return result
	}

	// Detect feed structure and parse accordingly
	switch v := rawJSON.(type) {
	case []interface{}:
		// Could be HackerNews (array of IDs) or Lobsters (array of objects)
		if len(v) > 0 {
			if _, ok := v[0].(float64); ok {
				// HackerNews: array of numeric IDs
				result = p.parseHackerNews(source, v)
			} else if _, ok := v[0].(map[string]interface{}); ok {
				// Lobsters: array of story objects
				result = p.parseLobsters(source, v)
			}
		}
	case map[string]interface{}:
		// Could be GitHub or Reddit (both have nested structure)
		if items, ok := v["items"].([]interface{}); ok {
			// GitHub: has "items" array
			result = p.parseGitHub(source, items)
		} else if data, ok := v["data"].(map[string]interface{}); ok {
			// Reddit: has "data" object
			if children, ok := data["children"].([]interface{}); ok {
				result = p.parseReddit(source, children)
			}
		}
	}

	if len(result.Items) == 0 && len(result.Errors) == 0 {
		result.Errors = append(result.Errors, "unrecognized feed structure")
	}

	return result
}

// parseHackerNews parses HackerNews top stories (array of IDs)
func (p *Parser) parseHackerNews(source string, items []interface{}) ParseResult {
	var result ParseResult

	for i, item := range items {
		id, ok := item.(float64)
		if !ok {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: expected numeric ID, got %T", i, item))
			continue
		}

		idStr := strconv.Itoa(int(id))
		title := fmt.Sprintf("HN Story %s", idStr)
		url := fmt.Sprintf("https://news.ycombinator.com/item?id=%s", idStr)

		feedItem := storage.FeedItem{
			ID:        p.generateID(source, url),
			Title:     title,
			URL:       url,
			Source:    source,
			CreatedAt: time.Now(),
		}

		result.Items = append(result.Items, feedItem)
	}

	return result
}

// parseGitHub parses GitHub API response
func (p *Parser) parseGitHub(source string, items []interface{}) ParseResult {
	var result ParseResult

	for i, item := range items {
		obj, ok := item.(map[string]interface{})
		if !ok {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: expected object, got %T", i, item))
			continue
		}

		// Extract required fields
		title, titleOk := p.getString(obj, "full_name")
		url, urlOk := p.getString(obj, "html_url")

		if !titleOk || !urlOk {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: missing required field (full_name or html_url)", i))
			continue
		}

		feedItem := storage.FeedItem{
			ID:        p.generateID(source, url),
			Title:     title,
			URL:       url,
			Source:    source,
			CreatedAt: time.Now(),
		}

		// Optional: timestamp
		if timestamp, ok := p.getString(obj, "updated_at"); ok {
			feedItem.Timestamp = &timestamp
		}

		// Optional: tags (topics)
		if topics, ok := obj["topics"].([]interface{}); ok {
			for _, topic := range topics {
				if topicStr, ok := topic.(string); ok {
					feedItem.Tags = append(feedItem.Tags, topicStr)
				}
			}
		}

		result.Items = append(result.Items, feedItem)
	}

	return result
}

// parseReddit parses Reddit API response
func (p *Parser) parseReddit(source string, children []interface{}) ParseResult {
	var result ParseResult

	for i, child := range children {
		childObj, ok := child.(map[string]interface{})
		if !ok {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: expected object, got %T", i, child))
			continue
		}

		data, ok := childObj["data"].(map[string]interface{})
		if !ok {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: missing data object", i))
			continue
		}

		// Extract required fields
		title, titleOk := p.getString(data, "title")
		url, urlOk := p.getString(data, "url")

		if !titleOk || !urlOk {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: missing required field (title or url)", i))
			continue
		}

		feedItem := storage.FeedItem{
			ID:        p.generateID(source, url),
			Title:     title,
			URL:       url,
			Source:    source,
			CreatedAt: time.Now(),
		}

		// Optional: timestamp (created_utc is Unix timestamp)
		if createdUtc, ok := data["created_utc"].(float64); ok {
			timestamp := time.Unix(int64(createdUtc), 0).Format(time.RFC3339)
			feedItem.Timestamp = &timestamp
		}

		// Optional: tags (link_flair_text)
		if flair, ok := p.getString(data, "link_flair_text"); ok && flair != "" {
			feedItem.Tags = []string{flair}
		}

		result.Items = append(result.Items, feedItem)
	}

	return result
}

// parseLobsters parses Lobsters API response
func (p *Parser) parseLobsters(source string, items []interface{}) ParseResult {
	var result ParseResult

	for i, item := range items {
		obj, ok := item.(map[string]interface{})
		if !ok {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: expected object, got %T", i, item))
			continue
		}

		// Extract required fields
		title, titleOk := p.getString(obj, "title")
		url, urlOk := p.getString(obj, "url")

		// Fall back to comments_url if url is not present
		if !urlOk {
			url, urlOk = p.getString(obj, "comments_url")
		}

		if !titleOk || !urlOk {
			result.Errors = append(result.Errors, fmt.Sprintf("item %d: missing required field (title or url)", i))
			continue
		}

		feedItem := storage.FeedItem{
			ID:        p.generateID(source, url),
			Title:     title,
			URL:       url,
			Source:    source,
			CreatedAt: time.Now(),
		}

		// Optional: timestamp
		if timestamp, ok := p.getString(obj, "created_at"); ok {
			feedItem.Timestamp = &timestamp
		}

		// Optional: tags
		if tags, ok := obj["tags"].([]interface{}); ok {
			for _, tag := range tags {
				if tagStr, ok := tag.(string); ok {
					feedItem.Tags = append(feedItem.Tags, tagStr)
				}
			}
		}

		result.Items = append(result.Items, feedItem)
	}

	return result
}

// getString attempts to extract a string value from a map
// It handles type coercion for common cases
func (p *Parser) getString(obj map[string]interface{}, key string) (string, bool) {
	val, ok := obj[key]
	if !ok {
		return "", false
	}

	switch v := val.(type) {
	case string:
		return v, true
	case float64:
		return strconv.FormatFloat(v, 'f', -1, 64), true
	case int:
		return strconv.Itoa(v), true
	case bool:
		return strconv.FormatBool(v), true
	default:
		// Try to JSON encode it as a fallback
		if bytes, err := json.Marshal(v); err == nil {
			return string(bytes), true
		}
		log.Printf("warning: cannot convert field %s to string (type: %T)", key, v)
		return "", false
	}
}

// generateID creates a deterministic ID from source name and URL
func (p *Parser) generateID(source, url string) string {
	hash := sha256.Sum256([]byte(source + url))
	return hex.EncodeToString(hash[:])
}
